File: me_daily/2025/11/07/lc_rank_474/b/maxProduct.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;
class Solution {
public:
    long long maxProduct(vector<int>& nums) {
        ll mx1 = 0, mx2 = 0;
        for (auto x : nums) {
            x = abs(x);
            if (x > mx1) {
                mx2 = mx1;
                mx1 = x;
            }else if (x >  mx2) {
                mx2 = x;
            }
        }
        return 1e5 * mx1 * mx2;
    }
};

File: me_daily/2025/11/07/lc_rank_474/b/maxProduct.py
def max_of_three(a: int, b: int, c: int) -> int:
    if a > b:
        if a > c:
            return a
    else:
        if b > c:
            return b
    return c

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        s = sorted(nums)
        p1 = s[0] * s[1]
        p2 = s[-1] * s[-2]
        p3 = s[0] * s[-1]
        max_two = max(p1, p2)
        min_two = min(p1, p2, p3)
        
        orig_max = max_of_three(
            s[0] * s[1] * s[-1],
            s[-3] * s[-2] * s[-1],
            0
        )
        
        cand1 = 100000 * max_two
        cand2 = -100000 * min_two
        
        return max_of_three(orig_max, cand1, cand2)